### PR TITLE
Plugins: Only set non-existing headers for core plugin requests

### LINKED ISSUE #78629: Plugins: New headers being overwritten by initial request headers
When an initial plugin request has headers set, if these headers are overwritten by some logic or middleware before the `HTTPClientMiddleware` is hit then the new values will always be replaced with the original values. 

The responsible code can be found [here](https://github.com/grafana/grafana/blob/710248674dfeaa5011b266a74b41f9f156511700/pkg/services/pluginsintegration/clientmiddleware/httpclient_middleware.go#L52). This is in contrast to the plugin sdk code which will only fill headers that do not already have a value (see [here](https://github.com/grafana/grafana-plugin-sdk-go/blob/v0.193.0/backend/data_adapter.go#L31)).

This can be validated by having a middleware that exchanges a users OAuth access token for a different value. When the final request is sent the original token will be present instead of the updated value.

### PR DESCRIPTION
**Pull request by aangelisc**
Fixes #78629 

When a plugin request has headers that have been set by some logic in the plugin or some other middleware this change prevents the `HTTPClient` middleware from overwriting those headers. This is the same logic as that which already exists in the plugin-sdk.

### CODE REVIEW

File: pkg/services/pluginsintegration/clientmiddleware/httpclient_middleware_test.go [hunk 1]
```diff
@@ -262,6 +262,36 @@ func TestHTTPClientMiddleware(t *testing.T) {
 				require.Len(t, middlewares, 1)
 				require.Equal(t, forwardPluginRequestHTTPHeaders, middlewares[0].(httpclient.MiddlewareName).MiddlewareName())
 
+				reqClone := req.Clone(req.Context())
+				// Create a header on the request as if it had been set by some other logic e.g. preceding middleware
+				reqClone.Header.Set(backend.OAuthIdentityTokenHeaderName, "bearer test-token")
```
Comment (andresmgot) [inline pkg/services/pluginsintegration/clientmiddleware/httpclient_middleware_test.go:267]: it makes sense to do this in the other test "Should not overwrite an existing header", no?

**Review by andresmgot — APPROVED** [review]
LGTM, makes sense.

**Review by wbrowne — COMMENTED** [review]
Please make sure to document this as a breaking change 🙏

**Comment by andresmgot** [comment]
> Please make sure to document this as a breaking change 🙏

Can this be considered a breaking change?

File: pkg/services/pluginsintegration/clientmiddleware/httpclient_middleware_test.go (hunk 1 above)
Comment (aangelisc) [reply pkg/services/pluginsintegration/clientmiddleware/httpclient_middleware_test.go:267]: Oops I put this in the wrong test! Will fix 😊 nice catch!

**Comment by wbrowne** [comment]
> > Please make sure to document this as a breaking change 🙏
> 
> Can this be considered a breaking change? I would say it's more of a bug fix.

Hmmm.. I guess depending on how you look at it, but not sure if someone could be tripped up by this as result and it'd be better to have notice like this to point to. I would say it's safer to add this, but I'd defer to @marefr as he has the most context here.

**Comment by marefr** [comment]
@aangelisc can you provide an example of headers causing problems for you? Might be security concerns to allow the client (end user) to override certain headers

**Comment by aangelisc** [comment]
@marefr the [Azure auth middleware](https://github.com/grafana/grafana-azure-sdk-go/blob/b20aac068928f54c7c89e32ef07b517d9cec8750/azhttpclient/middleware.go#L53) sets the `Authorization` header. When attempting to do user-based authentication the access token sent in the plugin request will be exchanged for another access token with the appropriate permissions. This middleware will then overwrite the new token with the original one which does not have the required permissions. 

Afaict, the plugin-sdk middleware will not overwrite a header if it has a non-empty value which is what I'd expect to happen in core as well.

**Review by marefr — APPROVED** [review]
LGTM and agree bugfix and not breaking and should only affect core plugins (I believe)

**Comment by aangelisc** [comment]
Thanks for the review comments all @andresmgot @wbrowne @marefr!

**Pull request merged**